### PR TITLE
Podfile Refactor

### PR DESCRIPTION
**Pull request by segiddins**
This part of the podfile refactor is pretty much done.

Closes https://github.com/CocoaPods/Core/issues/151.

- [x] Documentation
- [x] Sane commits
- [x] CHANGELOGs
- [x] More tests
- [x] Decide on friendly upgrade messages
- [x] Delete dead code

### CODE REVIEW

File: lib/cocoapods-core/podfile/dsl.rb [hunk 1]
```diff
@@ -277,28 +277,51 @@ def podspec(options = nil)
...
           raise Informative, "Unsupported options `#{options}` for " \
             "target `#{name}`"
         end
 
         parent = current_target_definition
         definition = TargetDefinition.new(name, parent)
-        definition.exclusive = true if options[:exclusive]
         self.current_target_definition = definition
-        yield
+        inherit!(:complete)
```
Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: shouldn't be necessary since its the default

File: lib/cocoapods-core/podfile/target_definition.rb [hunk 2]
```diff
@@ -13,6 +13,37 @@ class TargetDefinition
...
+        end
+        set_hash_value('inheritance', inheritance)
+      end
+
+      def inheritance
+        get_hash_value('inheritance', 'complete')
+      end
+
+      def installation_method=(installation_method)
+        set_hash_value('installation_method', installation_method)
```
Comment (segiddins) [inline lib/cocoapods-core/podfile/target_definition.rb]: validate this hash

File: lib/cocoapods-core/podfile/target_definition.rb [hunk 3]
```diff
@@ -368,7 +399,7 @@ def pod_whitelisted_for_configuration?(pod_name, configuration_name)
             end
           end
         end
-        !found && (exclusive? || parent.pod_whitelisted_for_configuration?(pod_name, configuration_name))
+        !found && (root? || (inheritance != 'none' && parent.pod_whitelisted_for_configuration?(pod_name, configuration_name)))
```
Comment (segiddins) [inline lib/cocoapods-core/podfile/target_definition.rb]: add a helper for `(root? || inheritance == 'none')`

**Comment by orta** [comment]
:tada:

**Comment by segiddins** [comment]
Preliminary docs added. @orta I'd greatly appreciate help making them more helpful for the guides

**Comment by segiddins** [comment]
Beyond the documentation bits, I think _this_ PR is ready for @CocoaPods/core to review.

**Comment by manuyavuz** [comment]
I'll try out today but that's huge!

File: spec/podfile_spec.rb [hunk 4]
```diff
@@ -114,6 +109,16 @@ module Pod
         Podfile.new { set_arc_compatibility_flag! }.should.set_arc_compatibility_flag
       end
 
+      it 'returns the installation method' do
+        name, options = Podfile.new {}.installation_method
+        name.should == 'cocoapods'
+        options.should == {}
```
Comment (orta) [inline spec/podfile_spec.rb]: I like this format ( https://github.com/artsy/eigen/pull/1003/files ) - I think it's good and future-proof :+1:

File: lib/cocoapods-core/podfile/dsl.rb [hunk 5]
```diff
@@ -273,32 +262,106 @@ def podspec(options = nil)
...
+      #
+      # @example Specifying custom CocoaPods installation options
+      #
+      #          install! 'cocoapods',
+      #                   :deterministic_uuids => false,
+      #                   :integrate_targets => false
+      #
+      # @return  [void]
+      #
+      def install!(installation_method, options = {})
```
Comment (orta) [inline lib/cocoapods-core/podfile/dsl.rb]: I wonder if this should be called `installation!` ( and ideally `installation` ) - the rest of the exposed CP API is about declaring what your integration should be, using a noun should feel better 

```
installation! 'cocoapods',
  :deterministic_uuids => false,
  :integrate_targets => false
```

File: lib/cocoapods-core/podfile/dsl.rb [hunk 6]
```diff
@@ -273,32 +262,106 @@ def podspec(options = nil)
...
+
+      # Sets the inheritance mode for the current target.
+      #
+      # @param   [:complete, :none, :search_paths] the inheritance mode to set.
+      #
+      # @example Inheriting only search paths
+      #
+      #          target 'App' do
+      #            target 'AppTests' do
+      #              inherit! :search_paths
```
Comment (orta) [inline lib/cocoapods-core/podfile/dsl.rb]: should `search_paths` be the default? This makes the pods from the parent target accessible to the child, which is what I'd expect to happen without any custom changes

**Comment by orta** [comment]
Might be the PR to remove these too:

```
module Pod
  class Specification
    module DSL
      module Deprecations
```

File: lib/cocoapods-core/podfile/dsl.rb (hunk 5 above)
Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: `use_frameworks!` is a verb, as is `ignore_all_warnings!`

File: lib/cocoapods-core/podfile/dsl.rb (hunk 6 above)
Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: the current default is analogous to `complete`, and I think it makes more sense given abstract targets now exist.

File: lib/cocoapods-core/podfile/dsl.rb [hunk 7]
```diff
@@ -273,32 +262,106 @@ def podspec(options = nil)
...
+      #             target :ZipAppTests do
       #               pod 'JSONKit'
       #             end
       #           end
       #
       # @return   [void]
       #
-      def target(name, options = {})
-        if options && !options.keys.all? { |key| [:exclusive].include?(key) }
+      def target(name, options = nil)
```
Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: > why have options at all? if we raise if they get sent at all?

for people who passed options in the current CocoaPods version. I'll add a TODO to remove the parameter, though, in a few months (it's just to make upgrading friendlier)

File: lib/cocoapods-core/podfile/dsl.rb (hunk 6 above)
Comment (orta) [inline lib/cocoapods-core/podfile/dsl.rb]: What does complete do differently? It links all of the pods

Comment (orta) [inline lib/cocoapods-core/podfile/dsl.rb]: Ideally the behaviour for what is probably the most common case ( like in Eigen, a big target then a unit test target subset ) shouldn't require the config change.

Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: yeah, complete is the same as how nested target definitions work now -- it inherits them exactly as if you'd specified them directly

File: lib/cocoapods-core/podfile/dsl.rb (hunk 7 above)
Comment (orta) [inline lib/cocoapods-core/podfile/dsl.rb]: legit :+1:

File: lib/cocoapods-core/podfile/dsl.rb (hunk 6 above)
Comment (neonichu) [inline lib/cocoapods-core/podfile/dsl.rb]: I think the default makes sense, as the most common case is probably sharing Pods between multiple targets.

File: lib/cocoapods-core/podfile/dsl.rb (hunk 5 above)
Comment (alloy) [inline lib/cocoapods-core/podfile/dsl.rb]: What are other possibilities for `installation_method` besides `cocoapods` ? And are these methods/options validated anywhere?

**Comment by alloy** [comment]
Great work on this @segiddins, nicely cleaned up :+1: :clap:

File: lib/cocoapods-core/podfile/dsl.rb (hunk 5 above)
Comment (segiddins) [inline lib/cocoapods-core/podfile/dsl.rb]: Right now, it's only `cocoapods`. But in a future version, plugins will be able to register installation methods. And the options are validated inside the `InstallationOptions` class in `CocoaPods/CocoaPods`

**Pull request merged**